// @inkline/inkline/components/IPopover/script.mjs
import { defineComponent } from "vue";
import {
  PopupMixin,
  PopupControlsMixin,
  sizePropValidator,
  colorVariantClass,
  defaultPropValue
} from "@inkline/inkline/mixins";
import ClickOutside from "@inkline/inkline/directives/click-outside";
import { uid } from "@inkline/inkline/helpers";
const componentName = "IPopover";
export default defineComponent({
  name: componentName,
  directives: {
    ClickOutside
  },
  mixins: [PopupMixin, PopupControlsMixin],
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    modelValue: {
      type: Boolean,
      default: false
    },
    name: {
      type: String,
      default() {
        return uid("popover");
      }
    },
    arrow: {
      type: Boolean,
      default: true
    },
    placement: {
      type: String,
      default: "top"
    },
    trigger: {
      type: [String, Array],
      default: () => ["click"]
    },
    offset: {
      type: Number,
      default: 6
    },
    interactable: {
      type: Boolean,
      default: false
    },
    popperOptions: {
      type: Object,
      default: () => ({})
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  emits: [
    "click-outside",
    "update:modelValue"
  ],
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size)
      };
    }
  },
  methods: {
    onEscape() {
      this.visible = false;
      this.$emit("update:modelValue", false);
    },
    handleClickOutside() {
      this.visible = false;
      this.$emit("update:modelValue", false);
      this.onClickOutside();
    }
  }
});
