// @inkline/inkline/components/IPopover/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

var _clickOutside = _interopRequireDefault(require("@inkline/inkline/directives/click-outside"));

var _helpers = require("@inkline/inkline/helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const componentName = "IPopover";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  directives: {
    ClickOutside: _clickOutside.default
  },
  mixins: [_mixins.PopupMixin, _mixins.PopupControlsMixin],
  props: {
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    modelValue: {
      type: Boolean,
      default: false
    },
    name: {
      type: String,

      default() {
        return (0, _helpers.uid)("popover");
      }

    },
    arrow: {
      type: Boolean,
      default: true
    },
    placement: {
      type: String,
      default: "top"
    },
    trigger: {
      type: [String, Array],
      default: () => ["click"]
    },
    offset: {
      type: Number,
      default: 6
    },
    interactable: {
      type: Boolean,
      default: false
    },
    popperOptions: {
      type: Object,
      default: () => ({})
    },
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    }
  },
  emits: ["click-outside", "update:modelValue"],
  computed: {
    classes() {
      return { ...(0, _mixins.colorVariantClass)(this),
        [`-${this.size}`]: Boolean(this.size)
      };
    }

  },
  methods: {
    onEscape() {
      this.visible = false;
      this.$emit("update:modelValue", false);
    },

    handleClickOutside() {
      this.visible = false;
      this.$emit("update:modelValue", false);
      this.onClickOutside();
    }

  }
});

module.exports = _default;
