// @inkline/inkline/components/ICheckbox/components/ICheckboxGroup/script.mjs
import { defineComponent } from "vue";
import {
  colorVariantClass,
  sizePropValidator,
  FormComponentMixin,
  defaultPropValue
} from "@inkline/inkline/mixins";
import { uid } from "@inkline/inkline/helpers";
const componentName = "ICheckboxGroup";
export default defineComponent({
  name: componentName,
  mixins: [
    FormComponentMixin
  ],
  provide() {
    return {
      formGroup: this
    };
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    inline: {
      type: Boolean,
      default: false
    },
    indeterminate: {
      type: Boolean,
      default: false
    },
    modelValue: {
      default: () => []
    },
    name: {
      type: [String, Number],
      default() {
        return uid("checkbox-group");
      }
    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  emits: [
    "update:modelValue"
  ],
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-disabled": this.isDisabled,
        "-readonly": this.isReadonly,
        "-inline": this.inline
      };
    },
    checked() {
      if (this.schema) {
        return this.schema.value;
      }
      return this.modelValue;
    }
  },
  methods: {
    onChange(value) {
      const modelValue = [...this.modelValue];
      const valueIndex = modelValue.findIndex((v) => v === value);
      if (valueIndex !== -1) {
        modelValue.splice(valueIndex, 1);
      } else {
        modelValue.push(value);
      }
      this.parent.onInput?.(this.name, modelValue);
      this.$emit("update:modelValue", modelValue);
    }
  }
});
