// @inkline/inkline/components/INavbar/script.mjs
import { defineComponent } from "vue";
import IContainer from "@inkline/inkline/components/IContainer/index.vue";
import IRow from "@inkline/inkline/components/IRow/index.vue";
import IColumn from "@inkline/inkline/components/IColumn/index.vue";
import IHamburgerMenu from "@inkline/inkline/components/IHamburgerMenu/index.vue";
import {
  CollapsibleMixin,
  defaultPropValue,
  colorVariantClass,
  sizePropValidator
} from "@inkline/inkline/mixins";
import { ClickOutside } from "@inkline/inkline/directives";
const componentName = "INavbar";
export default defineComponent({
  name: componentName,
  components: {
    IContainer,
    IRow,
    IColumn,
    IHamburgerMenu
  },
  directives: {
    ClickOutside
  },
  mixins: [
    CollapsibleMixin
  ],
  provide() {
    return {
      navbar: this
    };
  },
  props: {
    collapseOnItemClick: {
      type: Boolean,
      default: true
    },
    collapseOnClickOutside: {
      type: Boolean,
      default: true
    },
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    fluid: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    menuAnimation: {
      type: String,
      default: "close"
    }
  },
  emits: [
    "update:modelValue"
  ],
  computed: {
    classes() {
      return {
        ...this.collapsibleClasses,
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size)
      };
    }
  },
  methods: {
    onItemClick() {
      if (this.collapseOnItemClick && this.open) {
        this.setOpen(false);
      }
    },
    onClickOutside() {
      if (this.collapseOnClickOutside && this.open) {
        this.setOpen(false);
      }
    }
  }
});
