// @inkline/inkline/components/ITextarea/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _helpers = require("@inkline/inkline/helpers");

var _mixins = require("@inkline/inkline/mixins");

var _index = _interopRequireDefault(require("@inkline/inkline/components/IInput/index.vue"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const componentName = "ITextarea";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  extends: _index.default,
  props: {
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    clearable: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    id: {
      type: String,
      default: ""
    },
    modelValue: {
      type: String,
      default: ""
    },
    name: {
      type: [String, Number],

      default() {
        return (0, _helpers.uid)("textarea");
      }

    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    },
    tabindex: {
      type: [Number, String],
      default: 0
    }
  },
  emits: ["update:modelValue"]
});

module.exports = _default;
