// @inkline/inkline/components/ITextarea/script.mjs
import { defineComponent } from "vue";
import { uid } from "@inkline/inkline/helpers";
import {
  defaultPropValue,
  sizePropValidator
} from "@inkline/inkline/mixins";
import IInput from "@inkline/inkline/components/IInput/index.vue";
const componentName = "ITextarea";
export default defineComponent({
  name: componentName,
  extends: IInput,
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    clearable: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    id: {
      type: String,
      default: ""
    },
    modelValue: {
      type: String,
      default: ""
    },
    name: {
      type: [String, Number],
      default() {
        return uid("textarea");
      }
    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    tabindex: {
      type: [Number, String],
      default: 0
    }
  },
  emits: [
    "update:modelValue"
  ]
});
