// @inkline/inkline/components/IButtonGroup/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

const componentName = "IButtonGroup";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  inject: {
    form: {
      default: () => ({})
    },
    buttonGroup: {
      default: () => ({})
    },
    formGroup: {
      default: () => ({})
    }
  },

  provide() {
    return {
      buttonGroup: this
    };
  },

  props: {
    vertical: {
      type: Boolean,
      default: false
    },
    block: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      return {
        "-vertical": this.vertical,
        "-block": this.block,
        "-disabled": this.isDisabled
      };
    },

    isDisabled() {
      return this.disabled || this.buttonGroup.disabled || this.form.disabled || this.formGroup.disabled;
    }

  }
});

module.exports = _default;
