// @inkline/inkline/components/IDropdown/components/IDropdownItem/script.mjs
import { defineComponent } from "vue";
import {
  LinkableMixin
} from "@inkline/inkline/mixins";
const componentName = "IDropdownItem";
export default defineComponent({
  name: componentName,
  mixins: [
    LinkableMixin
  ],
  inject: {
    dropdown: {
      default: () => ({})
    }
  },
  props: {
    active: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    plaintext: {
      type: Boolean,
      default: false
    },
    tag: {
      type: String,
      default: "div"
    },
    tabindex: {
      type: [Number, String],
      default: 0
    }
  },
  computed: {
    classes() {
      return {
        "-active": this.active,
        "-disabled": this.disabled,
        "-plaintext": this.plaintext
      };
    },
    role() {
      return this.$attrs.to || this.$attrs.href ? "link" : "menuitem";
    },
    tabIndex() {
      return this.disabled ? -1 : this.tabindex;
    }
  },
  methods: {
    onClick(event) {
      this.dropdown.onItemClick?.(this, event);
    }
  }
});
