// @inkline/inkline/components/INumberInput/script.mjs
import { defineComponent } from "vue";
import IButton from "@inkline/inkline/components/IButton/index.vue";
import IInput from "@inkline/inkline/components/IInput/index.vue";
import { uid } from "@inkline/inkline/helpers";
import { defaultPropValue, sizePropValidator } from "@inkline/inkline/mixins";
const componentName = "INumberInput";
export default defineComponent({
  name: componentName,
  components: {
    IButton
  },
  extends: IInput,
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    clearable: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    id: {
      type: String,
      default: ""
    },
    modelValue: {
      type: [String, Number],
      default: ""
    },
    name: {
      type: [String, Number],
      default() {
        return uid("input");
      }
    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    tabindex: {
      type: [Number, String],
      default: 0
    },
    min: {
      type: [Number, String],
      default: -Infinity
    },
    max: {
      type: [Number, String],
      default: Infinity
    },
    precision: {
      type: Number,
      default: 0
    },
    step: {
      type: Number,
      default: 1
    }
  },
  emits: [
    "update:modelValue"
  ],
  watch: {
    modelValue: {
      immediate: true,
      handler(value) {
        let newValue = (value || "").toString().replace(/^[^0-9-]/, "").replace(/^(-)[^0-9]/, "$1").replace(new RegExp(`^(-?[0-9]+)[^0-9${this.precision > 0 ? "." : ""}]`), "$1");
        if (this.precision > 0) {
          newValue = newValue.replace(/^(-?[0-9]+\.)[^0-9]/, "$1").replace(new RegExp(`^(-?[0-9]+\\.[0-9]{0,${this.precision}}).*`), "$1");
        }
        if (parseFloat(newValue) >= parseFloat(this.max))
          newValue = this.max.toString();
        if (parseFloat(newValue) <= parseFloat(this.min))
          newValue = this.min.toString();
        this.parent.onInput?.(this.name, newValue);
        this.$emit("update:modelValue", newValue);
      }
    }
  },
  methods: {
    decrease() {
      this.$emit("update:modelValue", this.formatPrecision((Number(this.modelValue) - this.step).toString()));
    },
    increase() {
      this.$emit("update:modelValue", this.formatPrecision((Number(this.modelValue) + this.step).toString()));
    },
    formatPrecision(value) {
      const parts = value.split(".");
      let decimals = parts[1] || "";
      for (let i = decimals.length; i < this.precision; i += 1) {
        decimals += "0";
      }
      return this.precision > 0 ? `${parts[0]}.${decimals}` : parts[0];
    },
    onBlurFormatPrecision(event) {
      this.$emit("update:modelValue", this.formatPrecision(Number(this.modelValue).toString()));
      this.parent.onBlur?.(this.name, event);
    }
  }
});
