// @inkline/inkline/components/IHeader/script.mjs
import { defineComponent } from "vue";
import {
  defaultPropValue,
  sizePropValidator,
  colorVariantClass
} from "@inkline/inkline/mixins";
import IContainer from "@inkline/inkline/components/IContainer/index.vue";
import IRow from "@inkline/inkline/components/IRow/index.vue";
import IColumn from "@inkline/inkline/components/IColumn/index.vue";
const componentName = "IHeader";
export default defineComponent({
  name: componentName,
  components: {
    IContainer,
    IRow,
    IColumn
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    cover: {
      type: Boolean,
      default: false
    },
    fluid: {
      type: Boolean,
      default: false
    },
    fullscreen: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-cover": this.cover,
        "-fullscreen": this.fullscreen
      };
    }
  }
});
