// @inkline/inkline/components/INav/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

const componentName = "INav";

var _default = (0, _vue.defineComponent)({
  name: componentName,

  provide() {
    return {
      nav: this
    };
  },

  inject: {
    navbar: {
      default: () => ({
        onItemClick: () => {}
      })
    },
    sidebar: {
      default: () => ({
        onItemClick: () => {}
      })
    }
  },
  props: {
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    },
    vertical: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      return { ...(0, _mixins.colorVariantClass)(this),
        [`-${this.size}`]: Boolean(this.size),
        "-vertical": this.vertical
      };
    }

  },
  methods: {
    onItemClick() {
      [this.navbar, this.sidebar].forEach(parent => {
        parent.onItemClick();
      });
    }

  }
});

module.exports = _default;
