// @inkline/inkline/components/INav/script.mjs
import { defineComponent } from "vue";
import {
  defaultPropValue,
  colorVariantClass,
  sizePropValidator
} from "@inkline/inkline/mixins";
const componentName = "INav";
export default defineComponent({
  name: componentName,
  provide() {
    return {
      nav: this
    };
  },
  inject: {
    navbar: {
      default: () => ({
        onItemClick: () => {
        }
      })
    },
    sidebar: {
      default: () => ({
        onItemClick: () => {
        }
      })
    }
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    vertical: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-vertical": this.vertical
      };
    }
  },
  methods: {
    onItemClick() {
      [this.navbar, this.sidebar].forEach((parent) => {
        parent.onItemClick();
      });
    }
  }
});
