// @inkline/inkline/components/ICollapsible/components/ICollapsibleItem/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _helpers = require("@inkline/inkline/helpers");

var _transitions = require("@inkline/inkline/transitions");

const componentName = "ICollapsibleItem";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  components: {
    IExpandTransition: _transitions.IExpandTransition
  },
  inject: {
    collapsible: {
      default: () => ({
        activeItems: []
      })
    }
  },
  props: {
    name: {
      type: String,

      default() {
        return (0, _helpers.uid)("collapsible-item");
      }

    },
    title: {
      type: String,
      default: ""
    }
  },
  computed: {
    active() {
      return this.collapsible.activeItems.indexOf(this.name) > -1;
    },

    classes() {
      return {
        "-active": this.active
      };
    }

  },
  methods: {
    onClick() {
      this.collapsible.onItemClick(this);
    }

  }
});

module.exports = _default;
