// @inkline/inkline/components/IRadio/components/IRadioGroup/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

var _helpers = require("@inkline/inkline/helpers");

const componentName = "IRadioGroup";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  mixins: [_mixins.FormComponentMixin],

  provide() {
    return {
      formGroup: this
    };
  },

  props: {
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    inline: {
      type: Boolean,
      default: false
    },
    modelValue: {
      default: ""
    },
    name: {
      type: [String, Number],

      default() {
        return (0, _helpers.uid)("radio-group");
      }

    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    }
  },
  emits: ["update:modelValue"],
  computed: {
    classes() {
      return { ...(0, _mixins.colorVariantClass)(this),
        [`-${this.size}`]: Boolean(this.size),
        "-disabled": this.isDisabled,
        "-readonly": this.isReadonly,
        "-inline": this.inline
      };
    },

    checked() {
      if (this.schema) {
        return this.schema.value;
      }

      return this.modelValue;
    }

  },
  methods: {
    onChange(value) {
      this.parent.onInput?.(this.name, value);
      this.$emit("update:modelValue", value);
    }

  }
});

module.exports = _default;
