// @inkline/inkline/components/IRadio/components/IRadioGroup/script.mjs
import { defineComponent } from "vue";
import {
  colorVariantClass,
  sizePropValidator,
  FormComponentMixin,
  defaultPropValue
} from "@inkline/inkline/mixins";
import { uid } from "@inkline/inkline/helpers";
const componentName = "IRadioGroup";
export default defineComponent({
  name: componentName,
  mixins: [
    FormComponentMixin
  ],
  provide() {
    return {
      formGroup: this
    };
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    inline: {
      type: Boolean,
      default: false
    },
    modelValue: {
      default: ""
    },
    name: {
      type: [String, Number],
      default() {
        return uid("radio-group");
      }
    },
    readonly: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  emits: [
    "update:modelValue"
  ],
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-disabled": this.isDisabled,
        "-readonly": this.isReadonly,
        "-inline": this.inline
      };
    },
    checked() {
      if (this.schema) {
        return this.schema.value;
      }
      return this.modelValue;
    }
  },
  methods: {
    onChange(value) {
      this.parent.onInput?.(this.name, value);
      this.$emit("update:modelValue", value);
    }
  }
});
