// @inkline/inkline/components/ITabs/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

const componentName = "ITabs";

var _default = (0, _vue.defineComponent)({
  name: componentName,

  provide() {
    return {
      tabs: this
    };
  },

  props: {
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    modelValue: {
      type: String,
      default: ""
    },
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    },
    stretch: {
      type: Boolean,
      default: false
    }
  },
  emits: ["update:modelValue"],

  data() {
    return {
      active: this.modelValue,
      tabs: []
    };
  },

  computed: {
    classes() {
      return { ...(0, _mixins.colorVariantClass)(this),
        [`-${this.size}`]: Boolean(this.size),
        "-stretch": this.stretch
      };
    }

  },
  watch: {
    modelValue(value) {
      this.active = value;
    }

  },
  methods: {
    setActive(id) {
      this.active = id;
      this.$emit("update:modelValue", this.active);
    }

  }
});

module.exports = _default;
