// @inkline/inkline/components/ITabs/script.mjs
import { defineComponent } from "vue";
import {
  defaultPropValue,
  colorVariantClass,
  sizePropValidator
} from "@inkline/inkline/mixins";
const componentName = "ITabs";
export default defineComponent({
  name: componentName,
  provide() {
    return {
      tabs: this
    };
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    modelValue: {
      type: String,
      default: ""
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    stretch: {
      type: Boolean,
      default: false
    }
  },
  emits: [
    "update:modelValue"
  ],
  data() {
    return {
      active: this.modelValue,
      tabs: []
    };
  },
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-stretch": this.stretch
      };
    }
  },
  watch: {
    modelValue(value) {
      this.active = value;
    }
  },
  methods: {
    setActive(id) {
      this.active = id;
      this.$emit("update:modelValue", this.active);
    }
  }
});
