// @inkline/inkline/components/IButton/script.mjs
import { defineComponent } from "vue";
import ILoader from "@inkline/inkline/components/ILoader/index.vue";
import {
  LinkableMixin,
  sizePropValidator,
  colorVariantClass,
  defaultPropValue
} from "@inkline/inkline/mixins";
const componentName = "IButton";
export default defineComponent({
  name: componentName,
  components: {
    ILoader
  },
  mixins: [
    LinkableMixin
  ],
  inject: {
    buttonGroup: {
      default: () => ({})
    },
    form: {
      default: () => ({})
    },
    formGroup: {
      default: () => ({})
    }
  },
  props: {
    active: {
      type: Boolean,
      default: false
    },
    block: {
      type: Boolean,
      default: false
    },
    circle: {
      type: Boolean,
      default: false
    },
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    link: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    outline: {
      type: Boolean,
      default: false
    },
    tag: {
      type: String,
      default: "button"
    },
    tabindex: {
      type: [Number, String],
      default: 0
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  computed: {
    ariaBusy() {
      if (this.role !== "button") {
        return null;
      }
      return this.loading ? "true" : "false";
    },
    ariaDisabled() {
      if (this.role !== "button") {
        return null;
      }
      return this.disabled ? "true" : "false";
    },
    ariaPressed() {
      if (this.role !== "button") {
        return null;
      }
      return this.active ? "true" : "false";
    },
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-active": this.active,
        "-block": this.block,
        "-circle": this.circle,
        "-disabled": this.isDisabled,
        "-link": this.link,
        "-outline": this.outline
      };
    },
    isDisabled() {
      return this.disabled || this.buttonGroup.disabled || this.form.disabled || this.formGroup.disabled;
    },
    role() {
      return this.$attrs.to || this.$attrs.href ? "link" : "button";
    },
    tabIndex() {
      return this.isDisabled ? -1 : this.tabindex;
    }
  }
});
