// @inkline/inkline/components/INav/components/INavItem/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

const componentName = "INavItem";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  mixins: [_mixins.LinkableMixin],
  inject: {
    nav: {
      default: () => ({
        onItemClick: () => {}
      })
    }
  },
  inheritAttrs: false,
  props: {
    active: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    stopPropagation: {
      type: Boolean,
      default: false
    },
    tag: {
      type: String,
      default: "div"
    },
    tabindex: {
      type: [Number, String],
      default: 0
    }
  },
  computed: {
    ariaDisabled() {
      if (this.role === "link") {
        return null;
      }

      return this.disabled ? "true" : "false";
    },

    classes() {
      return {
        "-active": this.active,
        "-disabled": this.disabled
      };
    },

    role() {
      return this.$attrs.to || this.$attrs.href ? "link" : "menuitem";
    },

    tabIndex() {
      return this.disabled ? -1 : this.tabindex;
    }

  },
  methods: {
    onClick(event) {
      if (this.stopPropagation) {
        return;
      }

      this.nav.onItemClick(this, event);
    }

  }
});

module.exports = _default;
