// @inkline/inkline/components/IForm/components/IFormGroup/script.mjs
import { defineComponent } from "vue";
import {
  colorVariantClass,
  defaultPropValue,
  sizePropValidator,
  FormComponentMixin
} from "@inkline/inkline/mixins";
const componentName = "IFormGroup";
export default defineComponent({
  name: componentName,
  mixins: [
    FormComponentMixin
  ],
  provide() {
    return {
      formGroup: this
    };
  },
  props: {
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    disabled: {
      type: Boolean,
      default: false
    },
    inline: {
      type: Boolean,
      default: false
    },
    name: {
      type: String,
      default: ""
    },
    readonly: {
      type: Boolean,
      default: false
    },
    required: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    }
  },
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        [`-${this.size}`]: Boolean(this.size),
        "-disabled": this.isDisabled,
        "-readonly": this.isReadonly,
        "-inline": this.inline,
        "-required": this.required
      };
    }
  },
  methods: {
    onBlur(name, event) {
      this.parent?.onBlur?.(this.name ? `${this.name}.${name}` : name, event);
    },
    onInput(name, value) {
      this.parent?.onInput?.(this.name ? `${this.name}.${name}` : name, value);
    }
  }
});
