// @inkline/inkline/components/ITable/script.mjs
import { defineComponent } from "vue";
import { defaultPropValue, colorVariantClass } from "@inkline/inkline/mixins";
const componentName = "ITable";
export default defineComponent({
  name: componentName,
  props: {
    border: {
      type: Boolean,
      default: false
    },
    condensed: {
      type: Boolean,
      default: false
    },
    striped: {
      type: Boolean,
      default: false
    },
    hover: {
      type: Boolean,
      default: false
    },
    responsive: {
      type: [Boolean, String],
      default: true
    },
    nowrap: {
      type: Boolean,
      default: false
    },
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    }
  },
  computed: {
    classes() {
      return {
        ...colorVariantClass(this),
        "-border": this.border,
        "-condensed": this.condensed,
        "-striped": this.striped,
        "-hover": this.hover,
        "-nowrap": this.nowrap,
        [`-responsive${typeof this.responsive === "boolean" ? "" : `-${this.responsive}`}`]: Boolean(this.responsive)
      };
    }
  }
});
