// @inkline/inkline/components/ITable/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

const componentName = "ITable";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  props: {
    border: {
      type: Boolean,
      default: false
    },
    condensed: {
      type: Boolean,
      default: false
    },
    striped: {
      type: Boolean,
      default: false
    },
    hover: {
      type: Boolean,
      default: false
    },
    responsive: {
      type: [Boolean, String],
      default: true
    },
    nowrap: {
      type: Boolean,
      default: false
    },
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    }
  },
  computed: {
    classes() {
      return { ...(0, _mixins.colorVariantClass)(this),
        "-border": this.border,
        "-condensed": this.condensed,
        "-striped": this.striped,
        "-hover": this.hover,
        "-nowrap": this.nowrap,
        [`-responsive${typeof this.responsive === "boolean" ? "" : `-${this.responsive}`}`]: Boolean(this.responsive)
      };
    }

  }
});

module.exports = _default;
