// @inkline/inkline/components/IAlert/script.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
module.exports = void 0;

var _vue = require("vue");

var _mixins = require("@inkline/inkline/mixins");

const componentName = "IAlert";

var _default = (0, _vue.defineComponent)({
  name: componentName,
  props: {
    size: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "size"),
      validator: _mixins.sizePropValidator
    },
    color: {
      type: String,
      default: (0, _mixins.defaultPropValue)(componentName, "color")
    },
    modelValue: {
      type: Boolean,
      default: true
    },
    dismissible: {
      type: Boolean,
      default: false
    },
    dismissAriaLabel: {
      type: String,
      default: "Dismiss"
    }
  },
  emits: ["update:modelValue"],

  data() {
    return {
      dismissed: false
    };
  },

  computed: {
    classes() {
      return {
        [`-${this.color}`]: Boolean(this.color),
        [`-${this.size}`]: Boolean(this.size),
        "-dismissible": this.dismissible,
        "-with-icon": Boolean(this.$slots.icon)
      };
    }

  },
  watch: {
    modelValue(value) {
      this.dismissed = !value;
    }

  },
  methods: {
    dismiss() {
      this.dismissed = true;
      this.$emit("update:modelValue", false);
    }

  }
});

module.exports = _default;
