// @inkline/inkline/components/IAlert/script.mjs
import { defineComponent } from "vue";
import { defaultPropValue, sizePropValidator } from "@inkline/inkline/mixins";
const componentName = "IAlert";
export default defineComponent({
  name: componentName,
  props: {
    size: {
      type: String,
      default: defaultPropValue(componentName, "size"),
      validator: sizePropValidator
    },
    color: {
      type: String,
      default: defaultPropValue(componentName, "color")
    },
    modelValue: {
      type: Boolean,
      default: true
    },
    dismissible: {
      type: Boolean,
      default: false
    },
    dismissAriaLabel: {
      type: String,
      default: "Dismiss"
    }
  },
  emits: [
    "update:modelValue"
  ],
  data() {
    return {
      dismissed: false
    };
  },
  computed: {
    classes() {
      return {
        [`-${this.color}`]: Boolean(this.color),
        [`-${this.size}`]: Boolean(this.size),
        "-dismissible": this.dismissible,
        "-with-icon": Boolean(this.$slots.icon)
      };
    }
  },
  watch: {
    modelValue(value) {
      this.dismissed = !value;
    }
  },
  methods: {
    dismiss() {
      this.dismissed = true;
      this.$emit("update:modelValue", false);
    }
  }
});
