// @inkline/inkline/components/ISelect/components/ISelectOption/script.mjs
import { defineComponent } from "vue";
const componentName = "ISelectOption";
export default defineComponent({
  name: componentName,
  inject: {
    select: {
      default: () => ({})
    }
  },
  props: {
    active: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    label: {
      type: String,
      default: ""
    },
    tabindex: {
      type: [Number, String],
      default: 0
    },
    value: {
      type: [Object, String, Number],
      default: () => ({})
    }
  },
  computed: {
    ariaDisabled() {
      return this.disabled ? "true" : "false";
    },
    ariaSelected() {
      return this.active ? "true" : "false";
    },
    isActive() {
      return this.active || this.value === this.select.modelValue;
    },
    classes() {
      return {
        "-active": this.isActive,
        "-disabled": this.disabled
      };
    },
    tabIndex() {
      return this.disabled ? -1 : this.tabindex;
    }
  },
  methods: {
    onClick() {
      if (!this.disabled) {
        this.select.onInput(this.value, this.label);
      }
    }
  }
});
